File: src/chapters/numerical-calculus/content/code/simpson.cpp
#include <iostream>
#include <functional>
#include <cmath>
using namespace std;

// Composite Simpson's rule (n forced even) on [a, b].
double simpson(function<double(double)> f, double a, double b, int n = 100) {
    if (n % 2) ++n;
    double h = (b - a) / n, s = f(a) + f(b);
    for (int i = 1; i < n; ++i) s += (i % 2 ? 4 : 2) * f(a + i * h);
    return h / 3 * s;
}

int main() {
    cout.precision(10);
    cout << "int_0^1 e^x dx ~ " << simpson([](double x) { return exp(x); }, 0, 1, 100) << "\n";
}

File: src/pages/practice/guides/matlab/simpson.m
function I = simpson(f, a, b, n)
% SIMPSON  Composite Simpson's rule (n forced even) on [a,b].
    if nargin < 4, n = 100; end
    if mod(n, 2), n = n + 1; end
    x = linspace(a, b, n+1);
    y = f(x);
    h = (b - a) / n;
    I = h/3 * (y(1) + y(end) + 4*sum(y(2:2:end-1)) + 2*sum(y(3:2:end-1)));
end

% --- Demo ---
fprintf('int_0^1 e^x dx ~ %.10f\n', simpson(@exp, 0, 1, 100));
